datafusion/src/logical_plan/builder.rs

```diff
@@ -916,7 +916,11 @@ impl LogicalPlanBuilder {
 
     /// Build the plan
     pub fn build(&self) -> Result<LogicalPlan> {
-        Ok(self.plan.clone())
+        if let Some(err) = check_plan_invalid(&self.plan) {
```

Comment:
I recommend to return `Result` instead of `Option<Err>`.

We can use like this.

```rust
match err {
    Ok() => Ok(s),
    Err() => Err(e),
}
```

Reply:
You mean Result<LogicalPlan, DataFusionError>? 
Hm, this check function just finds the err, if it returns Ok(plan), it's a little bit strange🤔.
Another reason is we must write **return root_plan.clone()** in each arm of the match if use Result.
If you mean Result<(), DataFusionError>, I think it's the same as Option, right?
Or if I fail to get your idea?

Reply:
> If you mean Result<(), DataFusionError>, I think it's the same as Option, right?

Yes, A `result` represent either success/ Ok or failure/ Err. It's more elegant and reasonable.

Reply:
You can write this as well:

```rust
check_plan_invalid(&self.plan)?;
```

Which will return the error

Reply:
We need return the root plan instead of child plan.
If use `check_plan_invalid(&self.plan)?`, it cannot be recursively called like `check_plan_invalid(plan.input)?` because it returns child plan.